File: sdks/java/lib/src/main/java/com/hooksniff/api/MessagePoller.java
package com.hooksniff.api;

import com.hooksniff.HookSniffHttpClient;
import com.hooksniff.Utils;
import com.hooksniff.exceptions.ApiException;
import com.hooksniff.models.*;

import java.io.IOException;

public class MessagePoller {
    private final HookSniffHttpClient client;

    public MessagePoller(HookSniffHttpClient client) {
        this.client = client;
    }

    public MessagePollerPollResponse poll(String consumerId, Integer limit, String endpointId, String eventType, Boolean includePayload) throws IOException, ApiException {
        var builder = client.newUrlBuilder()
            .encodedPath("/api/v1/message-poller/poll")
            .queryParam("consumer_id", consumerId);
        if (limit != null) builder.queryParam("limit", limit.toString());
        if (endpointId != null) builder.queryParam("endpoint_id", endpointId);
        if (eventType != null) builder.queryParam("event_type", eventType);
        if (includePayload != null) builder.queryParam("include_payload", includePayload.toString());
        return client.executeRequest("GET", builder.build(), null, null, MessagePollerPollResponse.class);
    }

    public MessagePollerCursorResponse seek(String consumerId, String messageId, String endpointId) throws IOException, ApiException {
        var body = new java.util.HashMap<String, Object>();
        body.put("consumer_id", consumerId);
        body.put("message_id", messageId);
        if (endpointId != null) body.put("endpoint_id", endpointId);
        return client.executeRequest("POST", client.newUrlBuilder().encodedPath("/api/v1/message-poller/seek").build(), body, null, MessagePollerCursorResponse.class);
    }

    public MessagePollerCommitResponse commit(String consumerId, String messageId, String endpointId) throws IOException, ApiException {
        var body = new java.util.HashMap<String, Object>();
        body.put("consumer_id", consumerId);
        body.put("message_id", messageId);
        if (endpointId != null) body.put("endpoint_id", endpointId);
        return client.executeRequest("POST", client.newUrlBuilder().encodedPath("/api/v1/message-poller/commit").build(), body, null, MessagePollerCommitResponse.class);
    }
}

File: sdks/elixir/lib/hooksniff/resources/message_poller.ex
defmodule HookSniff.MessagePoller do
  alias HookSniff.Client

  @spec poll(Client.t(), String.t(), keyword()) :: {:ok, map()} | {:error, term()}
  def poll(%Client{} = c, consumer_id, opts \\ []) do
    params =
      [consumer_id: consumer_id, include_payload: Keyword.get(opts, :include_payload, true)]
      |> maybe_add(:limit, opts)
      |> maybe_add(:endpoint_id, opts)
      |> maybe_add(:event_type, opts)

    Client.request(c, :get, "/api/v1/message-poller/poll", nil, params)
  end

  @spec seek(Client.t(), String.t(), String.t(), keyword()) :: {:ok, map()} | {:error, term()}
  def seek(%Client{} = c, consumer_id, message_id, opts \\ []) do
    body =
      %{consumer_id: consumer_id, message_id: message_id}
      |> maybe_add_map(:endpoint_id, opts)

    Client.request(c, :post, "/api/v1/message-poller/seek", body)
  end

  @spec commit(Client.t(), String.t(), String.t(), keyword()) :: {:ok, map()} | {:error, term()}
  def commit(%Client{} = c, consumer_id, message_id, opts \\ []) do
    body =
      %{consumer_id: consumer_id, message_id: message_id}
      |> maybe_add_map(:endpoint_id, opts)

    Client.request(c, :post, "/api/v1/message-poller/commit", body)
  end

  defp maybe_add(params, key, opts) do
    case Keyword.get(opts, key) do
      nil -> params
      val -> Keyword.put(params, key, val)
    end
  end

  defp maybe_add_map(map, key, opts) do
    case Keyword.get(opts, key) do
      nil -> map
      val -> Map.put(map, key, val)
    end
  end
end
